File: lib/src/main/java/com/dhabensky/scopedvm/ScopedViewModelProviders.java
package com.dhabensky.scopedvm;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProvider.AndroidViewModelFactory;
import androidx.lifecycle.ViewModelProvider.Factory;
import androidx.lifecycle.ViewModelStore;
import androidx.lifecycle.ViewModelStoreOwner;

/**
 * Created on 30.03.2019.
 * @author dhabensky <dhabensky@yandex.ru>
 */
public class ScopedViewModelProviders {

	private final @NonNull ViewModelStoreOwner client;
	private final @Nullable String scope;

	private ScopedViewModelProviders(@NonNull ViewModelStoreOwner client, @Nullable String scope) {
		this.client = client;
		this.scope = scope;
	}

	public static ScopedViewModelProviders forScope(@NonNull Fragment fragment,
	                                                @Nullable String scope) {

		if (fragment.getLifecycle().getCurrentState() == Lifecycle.State.DESTROYED) {
			throw new IllegalArgumentException(
					"Cannot create ScopedViewModelProvider for destroyed client");
		}
		return new ScopedViewModelProviders(fragment, scope);
	}


	public ViewModelProvider of(@NonNull FragmentActivity activity) {
		Factory factory = AndroidViewModelFactory.getInstance(activity.getApplication());
		return of(activity, factory);
	}

	public ViewModelProvider of(@NonNull FragmentActivity activity, @NonNull Factory factory) {
		return ofStore(activity, factory);
	}

	public ViewModelProvider of(@NonNull Fragment fragment) {
		Factory factory = AndroidViewModelFactory
				.getInstance(fragment.getActivity().getApplication());
		return of(fragment, factory);
	}

	public ViewModelProvider of(@NonNull Fragment fragment, @NonNull Factory factory) {
		return ofStore(fragment, factory);
	}

	private ViewModelProvider ofStore(@NonNull ViewModelStoreOwner host, @NonNull Factory factory) {
		ViewModelStore store = Subscriptions.getScopedStore(host, scope, client);
		return new ViewModelProvider(store, factory);
	}

}

File: lib/src/main/java/com/dhabensky/svm/ScopedViewModelProviders.kt
package com.dhabensky.svm

import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider

/**
 * Created on 16.03.2019.
 * @author dhabensky <dhabensky@yandex.ru>
 */
class ScopedViewModelProviders
private constructor(
    private val requester: Fragment,
    private val scope: String?
) {

    companion object {
        @JvmStatic
        fun forScope(requester: Fragment, scope: String?): ScopedViewModelProviders {
            if (requester.lifecycle.currentState == Lifecycle.State.DESTROYED) {
                throw IllegalArgumentException("Cannot create ScopedViewModelProvider for destroyed requester")
            }
            return ScopedViewModelProviders(requester, scope)
        }
    }

    fun of(activity: FragmentActivity, factory: ViewModelProvider.Factory? = null)
            : ScopedViewModelProvider {

        val fact = factory ?: ViewModelProvider.AndroidViewModelFactory.getInstance(activity.application)

        return ScopedViewModelProvider(activity, fact, scope, requester)
    }

    fun of(fragment: Fragment, factory: ViewModelProvider.Factory? = null)
            : ScopedViewModelProvider {

        val fact = factory ?: ViewModelProvider.AndroidViewModelFactory.getInstance(fragment.activity!!.application)

        return ScopedViewModelProvider(fragment, fact, scope, requester)
    }

}
